Add support for pthread-based thread local storage

esteve opened the pull request:
This PR adds support for pthread-based thread local storage on platforms that don't have compiler builtin support for it. I only enabled it for iOS < 10.0, but it can be extended to any other platform via CMake flags.

esteve committed:
Added support for pthread-based thread local storage (e.g. for iOS < 10.0)

mikaelarguedas commented:
@ros2/team review this please

wjwwood commented on rmw/include/rmw/macros.h:
```diff
@@ -17,8 +17,21 @@
+  #endif
 #else
-  #define RMW_THREAD_LOCAL __thread
+  #define RMW_THREAD_LOCAL _Thread_local
```
Why this change for the non-iOS<10 and non-windows case?

esteve replied:
C11 added support for TLS via `_Thread_local`, `__thread` is a GCC extension that Clang supports for compatibility reasons. I can revert it to `__thread`, but I thought that given that IIRC ROS2 now requires a C11-capable compiler, it might be nice to use `_Thread_local` instead.

wjwwood replied:
The MSVC compiler is far from C11 compliant, AFAIK. So I think that we do not require a C11 compiler. It has been turned on in a few of the packages, but I had explicitly tried to avoid it in `rmw`. That being said, we already have to special case here for Windows, and the other compilers are likely C11 compliant. I'll either leave this as-is or try to detect when `_Thread_local` is not available and fallback further to `__thread`.

wjwwood commented on rmw/include/rmw/macros.h:
```diff
@@ -17,8 +17,21 @@
+    // iOS 10 added support for thread local storage
+    #if __IPHONE_OS_VERSION_MAX_ALLOWED < 100000
+      #define RMW_THREAD_LOCAL_PTHREAD 1
+      #define RMW_THREAD_LOCAL
```
So in this case the PTHREAD alternative is used?

esteve replied:
Exactly. If we happen to be targeting iOS < 10.0, the `pthread`-based implementation of TLS will be used (`RMW_THREAD_LOCAL_PTHREAD` is defined). This can be expanded to other platforms that have an implementation of `pthread`, but for now it's only for iOS < 10.0.

wjwwood replied:
After thinking about it, I think this might be somewhat dangerous because if someone is using `RMW_THREAD_LOCAL`, but doesn't special case it for `RMW_THREAD_LOCAL_PTHREAD` then they might end up with something that assumes it is thread local, but actually isn't because `RMW_THREAD_LOCAL` is just empty. I think I'll do `#undef RMW_THREAD_LOCAL` instead.

wjwwood reviewed (approved):
I had a few questions but otherwise it looks solid to me.

I'll start some CI.

wjwwood commented:
Oh, actually, it looks like it needs to be rebased. I'll do that too.

wjwwood commented on rmw/CMakeLists.txt:
```diff
@@ -21,13 +21,17 @@ set(rmw_sources
   ${rmw_sources}
   PROPERTIES LANGUAGE "C")
-add_library(${PROJECT_NAME} SHARED ${rmw_sources})
+add_library(${PROJECT_NAME} ${rmw_sources})
```
@dirk-thomas this is ok now that we have the "default to SHARED" turned on with https://github.com/ros2/ament_cmake_ros/blob/master/ament_cmake_ros/cmake/build_shared_libs.cmake, right?

dirk-thomas replied:
No, this package does currently not use `ament_cmake_ros` since #81 was reverted. Therefore the `SHARED` needs to stay.

wjwwood replied:
Should we just wait for #81 or should I try this change without removing `SHARED`? I guess it's not relevant to the rest of the thread local related changes.

@esteve I suppose that's up to you. I can easily remove this part of the patch and rerun CI for you.

esteve replied:
I've readded the `SHARED` keyword in https://github.com/ros2/rmw/pull/80/commits/911bb776da7f778c6ecfb1c36c3a0f27fa073949 until I can figure out why the `ament_cmake_ros` changes don't work.

wjwwood commented:
Looks like `rmw` is already defaulting to non-shared and causing issues:

```
02:10:57 /usr/bin/ld: /home/rosbuild/ci_scripts/ws/install/rmw/lib/librmw.a(error_handling.c.o): relocation R_X86_64_TPOFF32 against `__rmw_error_state' can not be used when making a shared object; recompile with -fPIC
```

-- http://ci.ros2.org/job/ci_linux/2410/consoleFull

@dirk-thomas what is blocking us from removing `SHARED` from the `add_library` call and still getting dynamic libraries by default?

@esteve since you're probably wanting to use this in the static linking mode at some point, I guess we should wait to merge this until the above issue is investigated?

wjwwood commented:
All of the CI I just ran, in case it's helpful for debugging:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2410)](http://ci.ros2.org/job/ci_linux/2410/)
- Linux-aarch64:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=46)](http://ci.ros2.org/job/ci_linux-aarch64/46/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1859)](http://ci.ros2.org/job/ci_osx/1859/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2529)](http://ci.ros2.org/job/ci_windows/2529/)

esteve committed:
Readd SHARED to add_library until ROS2 can be built statically via ament_cmake_ros

wjwwood commented:
Ok @esteve, here's new CI with your latest commit:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2418)](http://ci.ros2.org/job/ci_linux/2418/)
- Linux-aarch64:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=51)](http://ci.ros2.org/job/ci_linux-aarch64/51/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1871)](http://ci.ros2.org/job/ci_osx/1871/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2543)](http://ci.ros2.org/job/ci_windows/2543/)

wjwwood committed:
change detection of apple OS's

also undef RMW_THREAD_LOCAL to cause compiler
error in the case that pthread must be used

wjwwood committed:
line length fix

wjwwood committed:
remove SHARED again now that #98 has been merged

wjwwood commented:
CI:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2451)](http://ci.ros2.org/job/ci_linux/2451/)
- Linux-aarch64:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=75)](http://ci.ros2.org/job/ci_linux-aarch64/75/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1900)](http://ci.ros2.org/job/ci_osx/1900/) (known compiler warning)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2570)](http://ci.ros2.org/job/ci_windows/2570/)

wjwwood commented:
Thanks @esteve!

esteve commented:
@wjwwood thanks for fleshing out this PR. Just a minor comment, I remember deciding not using `TargetConditionals.h` (https://github.com/ros2/rmw/pull/80/commits/df1500bb35031bf4abd2504b3274c3f3ed4649c9), but unfortunately I forgot why. I think it was because `TargetConditionals.h` was not available in certain versions of the iOS SDK or because some of its definitions are deprecated, but nevertheless, if any issue shows up, it can be addressed in the future.

Thanks again for testing this!

wjwwood commented:
> @wjwwood thanks for fleshing out this PR. Just a minor comment, I remember deciding not using TargetConditionals.h (df1500b), but unfortunately I forgot why. I think it was because TargetConditionals.h was not available in certain versions of the iOS SDK or because some of its definitions are deprecated, but nevertheless, if any issue shows up, it can be addressed in the future.

Doh. That's what I get for trying to make it better. Ok we can address it if we need to in the future. Hopefully it works for some cases now.

> Thanks again for testing this!

No worries, I wanted to get it in before https://github.com/ros2/rmw/pull/101 because this whole code block is getting moved to another package for reuse in other places.

esteve commented:
@wjwwood oh sorry, I hope I didn't sound ungrateful, the code definitely looks much better now. It's actually my fault for not documenting why I didn't use `TargetConditionals.h`, now I forgot why I opted not to use it :smile:

wjwwood commented:
> @wjwwood oh sorry, I hope I didn't sound ungrateful

Not at all.

wjwwood commented:
@esteve do you know if pthread is still required for iOS? I thought I read that it's no longer required.

The reason I'm asking is that I'm refactoring the error handling code to no longer allocate memory when setting an error (trying to improve memory management) and the pthread specific code is making things really difficult.

I'm simultaneously looking into whether or not other potential platforms need pthread support still as well.

esteve commented:
I recall needing `pthread` for TLS with a Linaro toolchain a while ago, but I imagine that's no longer a problem. As for iOS, C11-based TLS is supported since 10. If this code needs to be revived in the future, I can take a look at it, but for now... KILL IT WITH FIRE

![](https://media.giphy.com/media/XvQXEi09zfmcE/giphy.gif)

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/announcing-objective-c-on-macos-and-ios-for-ros2/946/1
